<?xml version="1.0"?>
<launch>
    <!-- 修复版本的zhihang2025.launch - 解决spawn_model超时问题 -->
    
    <!-- 仿真参数 -->
    <arg name="est" default="ekf2"/>
    <arg name="world" default="$(find mavlink_sitl_gazebo)/worlds/zhihang2025.world"/>
    
    <!-- Gazebo配置 -->
    <arg name="gui" default="true"/>
    <arg name="debug" default="false"/>
    <arg name="verbose" default="false"/>
    <arg name="paused" default="false"/>
    <arg name="respawn_gazebo" default="true"/>
    <arg name="physics" default="ode"/>
    
    <!-- 启动Gazebo仿真环境 -->
    <include file="$(find gazebo_ros)/launch/empty_world.launch">
        <arg name="gui" value="$(arg gui)"/>
        <arg name="world_name" value="$(arg world)"/>
        <arg name="debug" value="$(arg debug)"/>
        <arg name="verbose" value="$(arg verbose)"/>
        <arg name="paused" value="$(arg paused)"/>
        <arg name="respawn_gazebo" value="$(arg respawn_gazebo)"/>
        <arg name="physics" value="$(arg physics)"/>
    </include>
    
    <!-- 第一架无人机: iris_0 -->
    <group ns="iris_0">
        <!-- 飞行器参数 -->
        <arg name="ID" value="0"/>
        <arg name="ID_in_group" value="0"/>
        <arg name="fcu_url" default="udp://:24540@localhost:34580"/>
        <arg name="vehicle" value="iris"/>
        <arg name="sdf" value="iris_zhihang"/>
        
        <!-- 位置参数 -->
        <arg name="x" value="2.5"/>
        <arg name="y" value="2.7"/>
        <arg name="z" value="1.3"/>
        <arg name="R" value="0"/>
        <arg name="P" value="0"/>
        <arg name="Y" value="0"/>
        
        <!-- 网络参数 -->
        <arg name="mavlink_udp_port" value="18570"/>
        <arg name="mavlink_tcp_port" value="4560"/>
        <arg name="udp_gimbal_port" value="13030"/>
        
        <!-- 生成模型描述参数 -->
        <param name="model_description" 
               command="xmlstarlet ed 
                       -d '//plugin[@name=&quot;gimbal_controller&quot;]/udp_gimbal_port_remote' 
                       -s '//plugin[@name=&quot;gimbal_controller&quot;]' -t elem -n udp_gimbal_port_remote -v $(arg udp_gimbal_port) 
                       -d '//plugin[@name=&quot;mavlink_interface&quot;]/mavlink_tcp_port' 
                       -s '//plugin[@name=&quot;mavlink_interface&quot;]' -t elem -n mavlink_tcp_port -v $(arg mavlink_tcp_port)
                       $(find px4)/Tools/sitl_gazebo/models/$(arg sdf)/$(arg sdf).sdf"/>
        
        <!-- PX4 SITL -->
        <node name="sitl_$(arg ID)" 
              pkg="px4" 
              type="px4" 
              output="screen" 
              args="$(find px4)/build/px4_sitl_default/etc -s etc/init.d-posix/rcS -i $(arg ID) -w sitl_$(arg vehicle)_$(arg ID)"
              respawn="true"
              respawn_delay="5"/>
        
        <!-- 延迟生成飞行器模型 -->
        <node name="$(arg vehicle)_$(arg ID)_spawn" 
              output="screen" 
              pkg="gazebo_ros" 
              type="spawn_model" 
              args="-sdf -param model_description -model $(arg vehicle)_$(arg ID_in_group) -x $(arg x) -y $(arg y) -z $(arg z) -R $(arg R) -P $(arg P) -Y $(arg Y)"
              launch-prefix="bash -c 'sleep 15; $0 $@'"
              respawn="false"/>
        
        <!-- MAVROS -->
        <include file="$(find mavros)/launch/px4.launch">
            <arg name="fcu_url" value="$(arg fcu_url)"/>
            <arg name="gcs_url" value=""/>
            <arg name="tgt_system" value="$(eval 1 + arg('ID'))"/>
            <arg name="tgt_component" value="1"/>
            <arg name="respawn_mavros" value="true"/>
        </include>
    </group>

    <!-- 第二架无人机: standard_vtol_0 -->
    <group ns="standard_vtol_0">
        <!-- 飞行器参数 -->
        <arg name="ID" value="1"/>
        <arg name="ID_in_group" value="0"/>
        <arg name="fcu_url" default="udp://:24541@localhost:34581"/>
        <arg name="vehicle" value="standard_vtol"/>
        <arg name="sdf" value="standard_vtol"/>
        
        <!-- 位置参数 -->
        <arg name="x" value="2.3"/>
        <arg name="y" value="0.4"/>
        <arg name="z" value="1.3"/>
        <arg name="R" value="0"/>
        <arg name="P" value="0"/>
        <arg name="Y" value="0"/>
        
        <!-- 网络参数 -->
        <arg name="mavlink_udp_port" value="18571"/>
        <arg name="mavlink_tcp_port" value="4561"/>
        <arg name="udp_gimbal_port" value="13031"/>
        
        <!-- 生成模型描述参数 -->
        <param name="model_description" 
               command="xmlstarlet ed 
                       -d '//plugin[@name=&quot;gimbal_controller&quot;]/udp_gimbal_port_remote' 
                       -s '//plugin[@name=&quot;gimbal_controller&quot;]' -t elem -n udp_gimbal_port_remote -v $(arg udp_gimbal_port) 
                       -d '//plugin[@name=&quot;mavlink_interface&quot;]/mavlink_tcp_port' 
                       -s '//plugin[@name=&quot;mavlink_interface&quot;]' -t elem -n mavlink_tcp_port -v $(arg mavlink_tcp_port)
                       $(find px4)/Tools/sitl_gazebo/models/$(arg sdf)/$(arg sdf).sdf"/>
        
        <!-- PX4 SITL -->
        <node name="sitl_$(arg ID)" 
              pkg="px4" 
              type="px4" 
              output="screen" 
              args="$(find px4)/build/px4_sitl_default/etc -s etc/init.d-posix/rcS -i $(arg ID) -w sitl_$(arg vehicle)_$(arg ID)"
              respawn="true"
              respawn_delay="5"/>
        
        <!-- 延迟生成飞行器模型 - 增加更长的延迟时间 -->
        <node name="$(arg vehicle)_$(arg ID)_spawn" 
              output="screen" 
              pkg="gazebo_ros" 
              type="spawn_model" 
              args="-sdf -param model_description -model $(arg vehicle)_$(arg ID_in_group) -x $(arg x) -y $(arg y) -z $(arg z) -R $(arg R) -P $(arg P) -Y $(arg Y)"
              launch-prefix="bash -c 'sleep 25; $0 $@'"
              respawn="false"/>
        
        <!-- MAVROS -->
        <include file="$(find mavros)/launch/px4.launch">
            <arg name="fcu_url" value="$(arg fcu_url)"/>
            <arg name="gcs_url" value=""/>
            <arg name="tgt_system" value="$(eval 1 + arg('ID'))"/>
            <arg name="tgt_component" value="1"/>
            <arg name="respawn_mavros" value="true"/>
        </include>
    </group>

</launch>
<!-- 修复版本的launch文件 - 添加了延迟启动和重启机制来解决spawn超时问题 -->
